bindings/rascal/models/krr.py

```diff
@@ -199,11 +232,78 @@ def get_representation_calculator(self):
         return self.kernel._rep
 
 
+# TODO(max, felix) I think this belongs in utils; it's not KRR-specific
+def get_grad_strides(frames):
+    """Get strides for total-energy/gradient kernels of the given structures
+
+    Parameters:
+        frames  List of structures each indicating the number of atoms
+
+    Returns: (1) the number of structures, (2) the number of gradient entries
+    (== 3 * the total number of atoms), and (3) strides for assigning the
+    gradient entries for each structure
+    """
+    Nstructures = len(frames)
+    Ngrad_stride = [0]
+    Ngrads = 0
+    for frame in frames:
+        n_at = len(frame)
+        Ngrad_stride.append(n_at * 3)
+        Ngrads += n_at * 3
+    Ngrad_stride = np.cumsum(Ngrad_stride) + Nstructures
+    return Nstructures, Ngrads, Ngrad_stride
+
+
+def compute_kernel_single(i_frame, frame, representation, X_sparse, kernel):
+    """Compute kernel of the (new) structure against the sparse points
+
+    Parameters:
+        i_frame     frame index (ignored???)
+        frame       New structure to compute kernel for
+        representation
+                    RepresentationCalculator to use for the structures
+        X_sparse    Sparse points to compute kernels against
+        kernel      Kernel object to use
+
+    Return both the kernel and the gradient of the kernel
+    """
+    feat = representation.transform([frame])
+    en_row = kernel(feat, X_sparse)
+    grad_rows = kernel(feat, X_sparse, grad=(True, False))
+    return en_row, grad_rows
+
+
+def compute_KNM(frames, X_sparse, kernel, soap):
+    """Compute kernel of the (new) structure against the sparse points
+
+    Parameters:
+        frame       New structure to compute kernel for
```

Comment:
the parameter is `frames` with an `s`. It looks like this function supports passing a list of frames, but the documentation disagree =)

Reply:
eh, typo. I had to add this documentation myself because the initial method wasn't documented :/